File: backend/app/routers/reports.py
from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import Session, select
from sqlalchemy import func
from datetime import datetime, date
from typing import Optional

from app.database import get_db
from app.models.pos_models import Transaction, TransactionItem, Product

router = APIRouter(prefix="/api/reports", tags=["Reports"])

# --- Helper Functions ---
def calculate_profit(session: Session, transactions):
    total_profit = 0.0
    for tx in transactions:
        tx_items = session.exec(
            select(TransactionItem).where(TransactionItem.transaction_id == tx.id)
        ).all()
        for item in tx_items:
            purchase_price = getattr(item, "purchase_price", getattr(item, "cost_price", 0.0))
            margin_per_unit = item.price - purchase_price
            total_profit += margin_per_unit * item.quantity
    return total_profit

def calculate_payment_breakdown(transactions):
    """Menghitung total nominal berdasarkan metode pembayaran (CASH, QRIS, TRANSFER, dll)"""
    breakdown = {}
    for tx in transactions:
        method = getattr(tx, "payment_method", "CASH") or "CASH"
        method = method.upper()
        total = getattr(tx, "grand_total", getattr(tx, "total", 0.0))
        breakdown[method] = breakdown.get(method, 0.0) + total
    return breakdown

# Helper function untuk format response transaksi
def format_transactions(transactions):
    result = []
    for tx in transactions:
        created_at_dt = getattr(tx, "created_at", None)
        
        date_str = "-"
        time_str = "-"
        iso_str = ""

        if isinstance(created_at_dt, datetime):
            date_str = created_at_dt.strftime("%d/%m/%Y")
            time_str = created_at_dt.strftime("%H:%M")
            iso_str = created_at_dt.isoformat()
        elif isinstance(created_at_dt, str):
            iso_str = created_at_dt
            if "T" in created_at_dt:
                parts = created_at_dt.split("T")
                date_str = datetime.strptime(parts[0], "%Y-%m-%d").strftime("%d/%m/%Y") if parts[0] else "-"
                time_str = parts[1][:5]
            elif " " in created_at_dt:
                parts = created_at_dt.split(" ")
                date_str = parts[0]
                time_str = parts[1][:5]

        invoice_no = getattr(tx, "invoice_number", getattr(tx, "invoice_no", f"INV-{tx.id}"))
        cashier = getattr(tx, "cashier_name", getattr(tx, "cashier", "Administrator"))

        result.append({
            "id": tx.id,
            "invoice_no": invoice_no,
            "created_at": iso_str,
            "date": date_str,
            "time": time_str,
            "cashier": cashier if cashier else "Administrator",
            "grand_total": getattr(tx, "grand_total", getattr(tx, "total", 0.0)),
            "payment_method": getattr(tx, "payment_method", "CASH")
        })
    return result

# 1. ENDPOINT LAPORAN HARIAN
@router.get("/daily")
@router.get("/daily-summary")
def get_daily_report(report_date: Optional[str] = None, session: Session = Depends(get_db)):
    target_date = datetime.strptime(report_date, "%Y-%m-%d").date() if report_date else date.today()

    transactions = session.exec(
        select(Transaction).where(func.date(Transaction.created_at) == str(target_date))
    ).all()

    total_sales = sum(getattr(tx, "grand_total", getattr(tx, "total", 0.0)) for tx in transactions)
    total_transactions = len(transactions)
    total_profit = calculate_profit(session, transactions)
    payment_breakdown = calculate_payment_breakdown(transactions)

    return {
        "summary": {
            "total_tx": total_transactions,
            "total_sales": total_sales,
            "total_profit": total_profit,
            "payment_breakdown": payment_breakdown
        },
        "transactions": format_transactions(transactions)
    }

# 2. ENDPOINT LAPORAN BULANAN
@router.get("/monthly")
def get_monthly_report(report_month: Optional[str] = None, session: Session = Depends(get_db)):
    if not report_month:
        report_month = date.today().strftime("%Y-%m")

    transactions = session.exec(
        select(Transaction).where(func.strftime("%Y-%m", Transaction.created_at) == report_month)
    ).all()

    total_sales = sum(getattr(tx, "grand_total", getattr(tx, "total", 0.0)) for tx in transactions)
    total_transactions = len(transactions)
    total_profit = calculate_profit(session, transactions)
    payment_breakdown = calculate_payment_breakdown(transactions)

    return {
        "summary": {
            "total_tx": total_transactions,
            "total_sales": total_sales,
            "total_profit": total_profit,
            "payment_breakdown": payment_breakdown
        },
        "transactions": format_transactions(transactions)
    }

# 3. ENDPOINT DETAIL TRANSAKSI / ITEM NOTA
@router.get("/transaction/{tx_id}")
def get_transaction_detail(tx_id: int, session: Session = Depends(get_db)):
    tx = session.get(Transaction, tx_id)
    if not tx:
        raise HTTPException(status_code=404, detail="Transaksi tidak ditemukan")

    items = session.exec(
        select(TransactionItem).where(TransactionItem.transaction_id == tx_id)
    ).all()

    item_list = []
    for it in items:
        product_name = getattr(it, "product_name", "Produk")
        if not product_name and getattr(it, "product_id", None):
            prod = session.get(Product, it.product_id)
            if prod:
                product_name = prod.name

        item_list.append({
            "product_name": product_name,
            "price": it.price,
            "quantity": it.quantity,
            "subtotal": it.price * it.quantity
        })

    created_at_str = tx.created_at.strftime("%d/%m/%Y %H:%M:%S") if getattr(tx, "created_at", None) else "-"

    return {
        "id": tx.id,
        "invoice_no": getattr(tx, "invoice_number", getattr(tx, "invoice_no", f"INV-{tx.id}")),
        "date_time": created_at_str,
        "cashier": getattr(tx, "cashier_name", getattr(tx, "cashier", "Administrator")),
        "payment_method": getattr(tx, "payment_method", "CASH"),
        "grand_total": getattr(tx, "grand_total", 0.0),
        "paid_amount": getattr(tx, "paid_amount", getattr(tx, "grand_total", 0.0)),
        "change_amount": getattr(tx, "change_amount", 0.0),
        "items": item_list
    }

File: frontend/web_client/js/reports.js
// Variable global untuk menyimpan data transaksi terakhir yang di-fetch
let currentReportTransactions = [];

// -------------------------------------------------------------
// GUARD ACCESS: Mencegah Kasir membuka laporan lewat URL langsung
// -------------------------------------------------------------
(function checkAdminAccess() {
  const currentUser = JSON.parse(localStorage.getItem("pos_current_user") || "{}");
  const role = (currentUser.role || "").toLowerCase();
  const isKasir = role === "kasir" || role === "cashier";

  if (!currentUser || !currentUser.role || isKasir) {
    alert("Akses Ditolak! Halaman Laporan Keuangan hanya dapat diakses oleh Admin.");
    window.location.href = "index.html";
  }
})();

document.addEventListener("DOMContentLoaded", () => {
  const reportTypeSelect = document.getElementById("report-type");
  const dailyFilter = document.getElementById("daily-filter");
  const monthlyFilter = document.getElementById("monthly-filter");
  const datePicker = document.getElementById("date-picker");
  const monthPicker = document.getElementById("month-picker");
  const btnLoadReport = document.getElementById("btn-load-report");

  // Inisialisasi tanggal & bulan default
  const today = new Date();
  const yyyy = today.getFullYear();
  const mm = String(today.getMonth() + 1).padStart(2, "0");
  const dd = String(today.getDate()).padStart(2, "0");

  if (datePicker) datePicker.value = `${yyyy}-${mm}-${dd}`;
  if (monthPicker) monthPicker.value = `${yyyy}-${mm}`;

  // Switch filter Harian / Bulanan
  if (reportTypeSelect) {
    reportTypeSelect.addEventListener("change", () => {
      if (reportTypeSelect.value === "daily") {
        if (dailyFilter) dailyFilter.style.display = "flex";
        if (monthlyFilter) monthlyFilter.style.display = "none";
      } else {
        if (dailyFilter) dailyFilter.style.display = "none";
        if (monthlyFilter) monthlyFilter.style.display = "flex";
      }
    });
  }

  if (btnLoadReport) {
    btnLoadReport.addEventListener("click", loadReportData);
  }

  // Listener tombol Export CSV
  const btnExportCSV = document.getElementById("btn-export-csv");
  if (btnExportCSV) {
    btnExportCSV.addEventListener("click", exportToCSV);
  }

  // Load data otomatis saat pertama kali dibuka
  loadReportData();
});

async function loadReportData() {
  const typeSelect = document.getElementById("report-type");
  const type = typeSelect ? typeSelect.value : "monthly";

  const dateVal = document.getElementById("date-picker")?.value;
  const monthVal = document.getElementById("month-picker")?.value;

  let endpoint = "";
  if (type === "daily") {
    endpoint = `/api/reports/daily?report_date=${dateVal}`;
  } else {
    endpoint = `/api/reports/monthly?report_month=${monthVal}`;
  }

  try {
    const response = await fetch(endpoint);
    if (!response.ok) {
      throw new Error(`HTTP Error Status: ${response.status}`);
    }

    const data = await response.json();

    // Simpan data transaksi ke variabel global untuk keperluan export
    currentReportTransactions = data.transactions || [];

    // 1. Render Tabel Transaksi
    renderReportTable(currentReportTransactions);

    // 2. Render KPI Cards & Breakdown Pembayaran
    renderKPI(
      data.summary || {
        total_tx: currentReportTransactions.length,
        total_sales: 0,
        total_profit: 0,
        payment_breakdown: {},
      },
    );
  } catch (error) {
    console.error("Gagal mengambil laporan dari API Backend:", error);
    alert("Terjadi kesalahan saat memuat laporan dari server.");
  }
}

function renderKPI(summary) {
  const statTx = document.getElementById("stat-total-tx");
  const statSales = document.getElementById("stat-total-sales");
  const statProfit = document.getElementById("stat-total-profit");

  if (statTx) statTx.innerText = summary.total_tx || 0;
  if (statSales) statSales.innerText = `Rp ${(summary.total_sales || 0).toLocaleString("id-ID")}`;
  if (statProfit) statProfit.innerText = `Rp ${(summary.total_profit || 0).toLocaleString("id-ID")}`;

  // Render breakdown metode pembayaran
  renderPaymentBreakdown(summary.payment_breakdown || {});
}

function renderPaymentBreakdown(breakdown) {
  const container = document.getElementById("payment-breakdown-container");
  if (!container) return;

  const defaultMethods = ["CASH", "QRIS", "DEBIT", "TRANSFER"];
  const colors = {
    CASH: "#10b981", // Hijau
    QRIS: "#3b82f6", // Biru
    DEBIT: "#f59e0b", // Oranye/Kuning
    TRANSFER: "#8b5cf6", // Ungu
  };

  const allMethods = {};
  defaultMethods.forEach((method) => {
    allMethods[method] = breakdown[method] || 0;
  });

  Object.keys(breakdown).forEach((method) => {
    if (!allMethods.hasOwnProperty(method)) {
      allMethods[method] = breakdown[method];
    }
  });

  // Inline style flex-direction row
  container.setAttribute("style", "display: flex !important; flex-direction: row !important; gap: 12px; margin-bottom: 20px; width: 100%;");

  container.innerHTML = Object.entries(allMethods)
    .map(([method, total]) => {
      const borderColor = colors[method] || "#6b7280";
      return `
        <div style="background: var(--bg-panel, #1f2937); border: 1px solid var(--border-color, #374151); border-left: 4px solid ${borderColor}; padding: 10px 14px; border-radius: 6px; flex: 1; min-width: 0;">
          <small style="color: var(--text-muted, #9ca3af); text-transform: uppercase; font-size: 11px; font-weight: bold; display: block;">METODE: ${method}</small>
          <div style="font-size: 15px; font-weight: bold; color: var(--text-main, #fff); margin-top: 3px; white-space: nowrap;">
            Rp ${(total || 0).toLocaleString("id-ID")}
          </div>
        </div>
      `;
    })
    .join("");
}

function renderReportTable(list) {
  const tbody = document.getElementById("report-table-body");
  if (!tbody) return;

  if (list.length === 0) {
    tbody.innerHTML = `<tr><td colspan="8" class="text-center" style="padding: 30px; color: #888;">Tidak ada data transaksi.</td></tr>`;
    return;
  }

  tbody.innerHTML = list
    .map((item, index) => {
      const invoice = item.invoice_no || `INV-${item.id}`;
      const cashier = item.cashier || "Administrator";
      const total = item.grand_total || 0;
      const dateDisplay = item.date || "-";
      const timeDisplay = item.time || "-";

      return `
            <tr>
                <td class="text-center">${index + 1}</td>
                <td><strong>${invoice}</strong></td>
                <td class="text-center">${dateDisplay}</td>
                <td class="text-center">${timeDisplay}</td>
                <td>${cashier}</td>
                <td class="text-center"><span class="badge-payment">${item.payment_method || "CASH"}</span></td>
                <td class="text-right"><strong>Rp ${total.toLocaleString("id-ID")}</strong></td>
                <td class="text-center">
                    <button class="btn-detail" onclick="openDetailModal(${item.id})">Detail</button>
                </td>
            </tr>
        `;
    })
    .join("");
}

// Buka Modal & Ambil Detail Transaksi
async function openDetailModal(txId) {
  try {
    const response = await fetch(`/api/reports/transaction/${txId}`);
    if (!response.ok) throw new Error("Gagal mengambil detail transaksi");

    const data = await response.json();

    document.getElementById("modal-invoice-no").innerText = data.invoice_no;
    document.getElementById("modal-date-time").innerText = data.date_time;
    document.getElementById("modal-cashier").innerText = data.cashier;
    document.getElementById("modal-payment-method").innerText = data.payment_method;
    document.getElementById("modal-grand-total").innerText = `Rp ${data.grand_total.toLocaleString("id-ID")}`;

    const itemsBody = document.getElementById("modal-items-body");
    itemsBody.innerHTML = data.items
      .map(
        (it) => `
            <tr>
                <td>${it.product_name}</td>
                <td class="text-center">${it.quantity}</td>
                <td class="text-right">Rp ${it.price.toLocaleString("id-ID")}</td>
                <td class="text-right">Rp ${it.subtotal.toLocaleString("id-ID")}</td>
            </tr>
        `,
      )
      .join("");

    document.getElementById("modal-detail").style.display = "flex";
  } catch (err) {
    console.error(err);
    alert("Gagal memuat detail nota.");
  }
}

function closeDetailModal() {
  document.getElementById("modal-detail").style.display = "none";
}

// Export Data Laporan ke File CSV
function exportToCSV() {
  if (!currentReportTransactions || currentReportTransactions.length === 0) {
    alert("Tidak ada data transaksi untuk diexport.");
    return;
  }

  const headers = ["No", "No Invoice", "Tanggal", "Jam", "Kasir", "Metode Pembayaran", "Total Transaksi"];

  const rows = currentReportTransactions.map((item, index) => {
    const invoice = `"${item.invoice_no || `INV-${item.id}`}"`;
    const date = `"${item.date || "-"}"`;
    const time = `"${item.time || "-"}"`;
    const cashier = `"${item.cashier || "Administrator"}"`;
    const payment = `"${item.payment_method || "CASH"}"`;
    const total = item.grand_total || 0;

    return [index + 1, invoice, date, time, cashier, payment, total];
  });

  let csvContent = "\uFEFF"; // Byte Order Mark (BOM) UTF-8
  csvContent += headers.join(",") + "\n";

  rows.forEach((row) => {
    csvContent += row.join(",") + "\n";
  });

  const blob = new Blob([csvContent], { type: "text/csv;charset=utf-8;" });
  const url = URL.createObjectURL(blob);

  const typeSelect = document.getElementById("report-type");
  const reportType = typeSelect ? typeSelect.value : "report";
  const dateVal = document.getElementById("date-picker")?.value;
  const monthVal = document.getElementById("month-picker")?.value;
  const period = reportType === "daily" ? dateVal : monthVal;

  const fileName = `Laporan_Penjualan_${period}.csv`;

  const link = document.createElement("a");
  link.setAttribute("href", url);
  link.setAttribute("download", fileName);
  link.style.visibility = "hidden";
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
}
